### PR TITLE
Update implemention of SlowMo to its implementation in Fairscale

### PR DESCRIPTION
**Pull request by vtantia**
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
SlowMo is being moved to [Fairscale](https://fairscale.readthedocs.io/en/latest/). This commit updates the implementation of SlowMo to the Fairscale version. It also adds tests for SlowMo.
Note: This PR is currently for review. It will be merged at a later date once SlowMo has been updated to Fairscale. SlowMo is being merged to Fairscale as part of [a PR](https://github.com/facebookresearch/fairscale/pull/378). So, once that PR is merged to Fairscale, this PR on Fairseq will be ready for merge

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairseq/models/distributed_fairseq_model.py
Comment (dianaml0) [inline fairseq/models/distributed_fairseq_model.py]: should `_GOSSIP_DISABLED` be updated to a different name?

File: fairseq/models/distributed_fairseq_model.py
Comment (dianaml0) [inline fairseq/models/distributed_fairseq_model.py:112]: Should we test that there is parity between `SlowMoDistributedDataParallel` and `GossipDataParallel` before merging?

**Review by dianaml0 — APPROVED** [review]
Just a few comments otherwise looks good to me!

File: fairseq/models/distributed_fairseq_model.py
Comment (vtantia) [reply fairseq/models/distributed_fairseq_model.py:112]: Good point. I'll explain a bit more. The changes involved in converting `GossipDataParallel` to `SlowMoDistributedDataParallel` are all refactors with no inside logic being changed. I've tested on Slurm on the WMT'16 En-De task using [these settings](https://github.com/pytorch/fairseq/blob/main/examples/scaling_nmt/README.md) that the accuracies remain exactly the same after switching these APIs, and they remain close to the accuracies of `DistributedDataParallel`. Do you have any other tests in mind, or would those be sufficient?

File: fairseq/models/distributed_fairseq_model.py
Comment (vtantia) [reply fairseq/models/distributed_fairseq_model.py]: Thanks have updated this

File: fairseq/models/distributed_fairseq_model.py
Comment (dianaml0) [reply fairseq/models/distributed_fairseq_model.py:112]: Thanks for testing it out! I think that should be sufficient.

**Comment by vtantia** [comment]
Thanks for the review! I'll wait for [SlowMo to be merged into Fairscale](https://github.com/facebookresearch/fairscale/pull/378) before going ahead with merging this PR.

**Pull request closed without merging**
